Cubical/Algebra/Matrix.agda

```diff
@@ -70,38 +82,128 @@ FinMatrix≡VecMatrix _ _ _ = ua FinMatrix≃VecMatrix
--- representation to the the other and relate the transported
--- operation with a more direct definition.
+-- Square matrices over a commuative ring form a commutative ring
 module _ (R : CommRing {ℓ}) where
 
   open commring-·syntax R
 
+  -- Operations
+
+  zeroFinMatrix : ∀ {m n} → FinMatrix ⟨ R ⟩ m n
+  zeroFinMatrix _ _ = ₀
+
+  negFinMatrix : ∀ {m n} → FinMatrix ⟨ R ⟩ m n → FinMatrix ⟨ R ⟩ m n
+  negFinMatrix M i j = - M i j
+
   addFinMatrix : ∀ {m n} → FinMatrix ⟨ R ⟩ m n → FinMatrix ⟨ R ⟩ m n → FinMatrix ⟨ R ⟩ m n
-  addFinMatrix M N = λ k l → M k l + N k l
+  addFinMatrix M N i j = M i j + N i j
+
+  oneFinMatrix : ∀ {n} → FinMatrix ⟨ R ⟩ n n
+  oneFinMatrix i j = if i == j then ₁ else ₀
+
+  sumVec : ∀ {n} → FinVec ⟨ R ⟩ n → ⟨ R ⟩
+  sumVec {zero} v  = ₀
+  sumVec {suc n} v = v zero + sumVec (v ∘ suc)
+
+  mulFinMatrix : ∀ {m1 m2 m3} → FinMatrix ⟨ R ⟩ m1 m2 → FinMatrix ⟨ R ⟩ m2 m3 → FinMatrix ⟨ R ⟩ m1 m3
+  mulFinMatrix M N i k = sumVec λ j → M i j · N j k
+
+  -- Properties
+
+  isSetFinMatrix : ∀ {m n} → isSet (FinMatrix ⟨ R ⟩ m n)
+  isSetFinMatrix = isSetΠ2 λ _ _ → commring-is-set R
+
+  addFinMatrixAssoc : ∀ {m n} → (M N K : FinMatrix ⟨ R ⟩ m n) → addFinMatrix M (addFinMatrix N K) ≡ addFinMatrix (addFinMatrix M N) K
+  addFinMatrixAssoc M N K i j k = commring+-assoc R (M j k) (N j k) (K j k) i
+
+  addFinMatrix0r : ∀ {m n} → (M : FinMatrix ⟨ R ⟩ m n) → addFinMatrix M zeroFinMatrix ≡ M
+  addFinMatrix0r M i j k = commring+-rid R (M j k) i
+
+  addFinMatrix0l : ∀ {m n} → (M : FinMatrix ⟨ R ⟩ m n) → addFinMatrix zeroFinMatrix M ≡ M
+  addFinMatrix0l M i j k = commring+-lid R (M j k) i
+
+  addFinMatrixNegMatrixr : ∀ {m n} → (M : FinMatrix ⟨ R ⟩ m n) → addFinMatrix M (negFinMatrix M) ≡ zeroFinMatrix
+  addFinMatrixNegMatrixr M i j k = commring+-rinv R (M j k) i
+
+  addFinMatrixNegMatrixl : ∀ {m n} → (M : FinMatrix ⟨ R ⟩ m n) → addFinMatrix (negFinMatrix M) M ≡ zeroFinMatrix
+  addFinMatrixNegMatrixl M i j k = commring+-linv R (M j k) i
 
   addFinMatrixComm : ∀ {m n} → (M N : FinMatrix ⟨ R ⟩ m n) → addFinMatrix M N ≡ addFinMatrix N M
   addFinMatrixComm M N i k l = commring+-comm R (M k l) (N k l) i
 
+  mulFinMatrixAssoc : ∀ {m n k l} → (M : FinMatrix ⟨ R ⟩ m n) → (N : FinMatrix ⟨ R ⟩ n k) → (K : FinMatrix ⟨ R ⟩ k l)
+                   → mulFinMatrix M (mulFinMatrix N K) ≡ mulFinMatrix (mulFinMatrix M N) K
+  mulFinMatrixAssoc M N K i j k = {!!}
+
+  mulFinMatrix1r : ∀ {m n} → (M : FinMatrix ⟨ R ⟩ m n) → mulFinMatrix M oneFinMatrix ≡ M
+  mulFinMatrix1r {n = zero} M = funExt₂ λ j → \ { () }
+  mulFinMatrix1r {n = suc n} M = funExt₂ λ j k → {!!}
+
+  mulFinMatrix1l : ∀ {m n} → (M : FinMatrix ⟨ R ⟩ m n) → mulFinMatrix oneFinMatrix M ≡ M
+  mulFinMatrix1l M i j k = {!!}
+
+  mulFinMatrixComm : ∀ {n} → (M N : FinMatrix ⟨ R ⟩ n n) → mulFinMatrix M N ≡ mulFinMatrix N M
+  -- Agda bug?
+  mulFinMatrixComm {n = n} M N i j k = sumVec {n = {!n!}} (λ l → commring·-comm R (M j k) (N k j) i)
```

Comment:
@Saizan I think I might be hitting some bug here... If I try to input `n` in the hole I get the error:
```
x₁ != x of type Fin n
when checking the definition of mulFinMatrixComm
```

There's now `x` around so I have no clue what is happening. In some other places I also get weird errors where I haven't introduced any `n₁` and I would expect to be able to provide `i` but Agda doesn't like it (despite the mysterious  `n₁` being equal to `n`):
```
Goal: Fin n₁
————————————————————————————————————————————————————————————
j      : Fin n
k      : Fin n
i      : Fin n
n      : ℕ
M      : FinMatrix ⟨ R ⟩ n₁ n₁
n = n₁ : ℕ   (not in scope)
```

Reply:
I can't reproduce the weirdness about `n`, but there's a legitimate reason not to accept that proof. The error message mentioning `x` however is an error reporting bug caused by the internal machinery losing track of the user-generated names.

With the following code

```agda
  mulFinMatrixComm : ∀ {n} → (M N : FinMatrix ⟨ R ⟩ n n) → mulFinMatrix M N ≡ mulFinMatrix N M
  mulFinMatrixComm {n = n} M N i j k = sumVec {n = n} \ l → { }6
```

I get the following constraints for the goal:

```agda
———— Constraints ———————————————————————————————————————————
R .snd .fst .snd .snd (M j l) (N l k) = ?6 (i = i0) : R .fst
R .snd .fst .snd .snd (N j l) (M l k) = ?6 (i = i1) : R .fst
```

notice that while `M` and `N` switch places they also get applied to different indexes, so just commutativity in the `R` ring is not enough to prove this.

Reply:
Oh I see, my bad! I guess Agda is smarter than me sometimes... Is the bug related to losing track of user-generated names reported? I would probably have realized what was wrong if the error message wasn't so weird...

Reply:
I don't think it's reported, would be nice to have a small testcase.